pkg/helpers/helpers.go

```diff
@@ -33,3 +33,19 @@ func IsObjectTaggedToDelete(obj client.Object) bool {
 	annotation, ok := annotations[DeleteTagAnnotation]
 	return ok && annotation == "true"
 }
+
+func MergeMapStringString(modified *bool, existing map[string]string, desired *map[string]string) {
+	if existing == nil {
+		existing = map[string]string{}
+	}
+
+	// for each desired key value set, e.g. labels
+	// check if it's present in existing. if not add it to existing.
+	// e.g. preserving existing labels while adding those that are in the desired set.
+	for desiredKey, desiredValue := range *desired {
+		if existingValue, exists := (existing)[desiredKey]; !exists || existingValue != desiredValue {
+			(*desired)[desiredKey] = existingValue
+			*modified = true
+		}
+	}
```

Comment:
nit: [Go maps are by default passed by reference](https://go.dev/blog/maps), I don't think we need deference `desired *map[string]string` typically? Would make the function a bit simpler 🤔 
```suggestion
func MergeMapStringString(modified *bool, existing map[string]string, desired map[string]string) {
	if existing == nil {
		existing = map[string]string{}
	}

	// for each desired key value set, e.g. labels
	// check if it's present in existing. if not add it to existing.
	// e.g. preserving existing labels while adding those that are in the desired set.
	for desiredKey, desiredValue := range desired {
		if existingValue, exists := existing[desiredKey]; !exists || existingValue != desiredValue {
			desired[desiredKey] = existingValue
			*modified = true
		}
	}
```

Reply:
If you check my proposal, it is the _existing_ map which is passed as a pointer, because it could be empty and it is being initialized within the function. The _desired_ map is _read-only_ so, no need to be passed as a pointer. a reference is enough.

Reply:
And it is the _existing_ map which needs to be updated. The _desired_ represents an idea, a willing. The actual thing is the existing.

Reply:
Thanks @eguzki I though I had addressed it, i updated the comment but not the assignment. Should be good now. 👍

Reply:
Oh right, but I think my nit still applies to how the existing map is passed. I've ran through the verifications steps with the function using `existing map[string]string` instead on `*existing map[string]string` and the map was still updated correctly 🤔

Reply:
It might be that the _existing_ map is not empty or it is empty and initialized. This method even accept `nil` value for object which type is `existing *map[string]string`. The pointer is only for enabling 
```go
if *existing == nil {                                                                         
        *existing = map[string]string{}                                                           
}
```

As you rightly pointed out, there is no need to pass a pointer to update in place a map because the maps are passed as references.

Reply:
However, I guess it's possible to get a `nil` pointer deference if `nil` is passed at `if *existing == nil`? 🤔 
For example:
```go
func TestMergeMapStringString(t *testing.T) {
	type args struct {
		modified *bool
		existing *map[string]string
		desired  map[string]string
	}
	tests := []struct {
		name string
		args args
	}{
		{
			name: "nil modified",
			args: args{
				modified: ptr.To(false),
				existing: nil,
				desired:  map[string]string{"test": "test"},
			},
		},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			MergeMapStringString(tt.args.modified, tt.args.existing, tt.args.desired)
			assert.Assert(t, *tt.args.modified == true)
			assert.Assert(t, reflect.DeepEqual(*tt.args.existing, tt.args.desired))
		})
	}
}
```
Doesn't happen when used in the context of `DeploymentTemplateLabelMutator` but maybe some other function that uses `MergeMapStringString` directly in the future encounter this 🤷 

Anyhow, I'll leave this here on whether this should be addressed or not

Reply:
Ah, the beauty of the double indirection because maps are references and we need it to initialize in case it is `nil`. 

Let me explain the goal for having that signature
```go
func MergeMapStringString(modified *bool, existing *map[string]string, desired map[string]string)
```

We want that type to be a pointer to a map because that method will be used, among other use cases as you pointed out, like the following:

```go
MergeMapStringString(&updated, &existing.Spec.Template.Labels, desired.Spec.Template.Labels)
```

It's one liner to ensure desired labels will be in the existing objects labels field. Note that the map is one field of a struct type (the k8s obbject) holding the map, in this example: `existing.Spec.Template.Labels`. That field is usually initialized from a deserialization (unmarshalling) process from a network object coming from k8s API server. It is very common that k8s objects do not have labels, so that field's value is `nil`, not empty. However, the `MergeMapStringString` will merge desired values to the existing ones even if this existing one is `nil` (which is different from empty). So, the method will get the address of the field holding the map (which is a reference, hence the double indirection fact here) and can initialize on the object itself.

If the method signature were 
```go
func MergeMapStringString(modified *bool, existing map[string]string, desired map[string]string)
```

The method could not be used to populate a k8s object with nil map. 

Anyway, good catch, the function missed one guard to protect itself from uses cases other than that explained above. 

I have written the fix and unittests:

```go
func MergeMapStringString(modified *bool, existing *map[string]string, desired map[string]string) {

	if existing == nil {
		// early return, as there is no ref to initialize map
		return
	}

	if *existing == nil {
		*existing = map[string]string{}
	}

	for k, v := range desired {
		if existingVal, ok := (*existing)[k]; !ok || v != existingVal {
			(*existing)[k] = v
			*modified = true
		}
	}
}
```

```go
func TestMergeMapStringString(t *testing.T) {
	type args struct {
		modified *bool
		existing map[string]string
		desired  map[string]string
	}
	tests := []struct {
		name string
		args args
	}{
		{
			name: "nil map",
			args: args{
				modified: ptr.To(false),
				existing: nil,
				desired:  map[string]string{"test": "test"},
			},
		},
		{
			name: "empty map",
			args: args{
				modified: ptr.To(false),
				existing: make(map[string]string),
				desired:  map[string]string{"test": "test"},
			},
		},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			MergeMapStringString(tt.args.modified, &tt.args.existing, tt.args.desired)
			assert.Assert(t, *tt.args.modified) // modified
			assert.Assert(t, reflect.DeepEqual(tt.args.existing, tt.args.desired))
		})
	}

	t.Run("nil pointer", func(t *testing.T) {
		desired := map[string]string{"test": "test"}
		modified := false
		MergeMapStringString(&modified, nil, desired)
		assert.Assert(t, !modified) // not modified
	})
}
```

If you happen to know a different way to achieve that, I am very happy to learn. My golang is getting rusty in favor of Rust :)

Reply:
Ah, yes, I see, that makes sense. Looks good to me 👍